# Final exam

In [1]:
# Load packages:

# this package allows to work efficiently with arrays
import numpy as np
# this package provides some tools of linear algebra. We can use them to simplify certain algorithms and for comparisons.
from scipy.linalg import norm

---

**NAME:**

---

**Instructions before starting:** 
- The materials from the lecture and the previous tutorial are allowed. But <font color=red>you are **not allowed to communicate or to exchange files**</font> (Moodle uses tools to prevent from plagiarism). 
- The **clarity** and the organization of the work are taken into account during the correction. 
- Before uploading your notebook on moodle, do **Kernel/restart and clear output**. And **verify that your code is running cell after cell.**
- The notebook is long and it is **not necessary to finish all the questions** to obtain optimal grade. 
- The theoretical questions should be answered **in the notebook**! In order to help you typing mathematical formulae, you will find a few commands in the next cell some that you can copy-paste. 
- The notebook should be uploaded back on moodle **before 10**! Take your precautions. 
- Before uploading the notebook, do **Kernel/Restart and clear output**. Your code should run cell after cell. 

---

(Double click on this cell to edit it)
 
**Commands for mathematical formulae:**
- Open a formulae $math formula$ (between dolars) or long formulae $$math formula$$ (between double dollars)
- Fraction $$\frac{a}{b}$$
- Square root $$\sqrt{\frac{a}{b}}$$
- Open and close a long parenthesis $$\left(\sqrt{\frac{a}{b}} \right)$$
- Indices: down $A_{i,j,k}$ (between brackets for several letters) or top $A^{i,j,k}$
- Matrices: a matrix is an array between parenthesis. Arrays required a second argument, that is the number of columns and how to align the formula within a column (c=center, l=left, r=right). Every column component is separated by a \& and every line is separated by two '\'. For example, 
    $$A = \left( \begin{array}{cc} 1 & 2 \\ 3 & 4\end{array}\right), \qquad
    B = \left( \begin{array}{ccc} 1 & 2 & 8 \\ 3 & 4 & 2 \end{array}\right), \qquad 
    D = \left( \begin{array}{cc} 1 & 2 \\ 3 & 4 \\ 5 &  6\end{array}\right) $$
    You can also define it componentwise $A_{i,j} = i + j$ for $i=1,\dots,N$ and $j=1,\dots,N$ 
- Align equations with  
  \begin{align} 
  a &= b \\ 
  &=c \\ 
  &\neq d
  \end{align} 
  
(ctrl + Enter to run it)

---

---

## Exercise

Consider the matrix 

$$ A = \left( \begin{array}{cccccc} 
0 & 0 & a & b & 0 & 0 \\
0 & 0 & c & d & 0 & 0 \\
e & f & 0 & 0 & 0 & 0 \\ 
g & h & 0 & 0 & 0 & 0 \\
0 & 0 & 0 & 0 & i & j \\
0 & 0 & 0 & 0 & k & l
\end{array}\right) $$

1) Compute its determinant. 

*Hints:* You may introduce intermediate determinants that can also be used in the next question.

**Answer:**
We devide $A$ into 9 parts. Denote $A_1 = A_3 = A_5 = A_6 = A_7 = A_8 = \left( \begin{array}{cc} 0 & 0  \\ 0 & 0 \end{array}\right), A_2 = \left( \begin{array}{cc} a & b  \\ c & d \end{array}\right), A_4 = \left( \begin{array}{cc} e & f  \\ g & h \end{array}\right), A_9 = \left( \begin{array}{cc} i & j  \\ k & l \end{array}\right)$.

We have $det(A_i) = 0 \quad \forall i \in \{1, 3,5,6,7,8\}.$

Then $det(A) =  det(A_2) * det(A_4) * det(A_9) = -(ad -bc)(eh-gf)(il-jk)$.

2) Assuming $det(A)\neq 0$, use Cramer rule to compute its inverse. Explain the simplifications. 

**Answer:** 

We compute the comatrix of $A$ based on the previous devided $A$. We know that $Com(A)$ does not zero entry only if it require to compute the product of two non-zero matrices, i.e. at position of $A_2$, $A_4$ and $A_9$. Then we computye these 12 non-zero entries individually. 

$Com(A)_{1, 3} = d * det(A_4) * det(A_9)$
$Com(A)_{1, 4} = -c * det(A_4) * det(A_9)$

We found the pattern: for entries in position A_2, the entry is the diagonal position in A_2, times the sign of this entry according to pairity, and times determinant of A_4 and that of A_9. 

We take the transpose of comatrix and get


$$A^{-1} = \frac{1}{-(ad -bc)(eh-gf)(il-jk)} \left(\begin{array}{cccccc} 
    0      & 0      & d(eh-gf)(il-jk)       & -c(eh-gf)(il-jk)      & 0      & 0     \\
    0      & 0      & -b(eh-gf)(il-jk)      & a(eh-gf)(il-jk)      & 0      & 0     \\
    h(ad -bc)(il-jk)      & -g(ad -bc)(il-jk)      & 0      & 0      & 0      & 0     \\
    -f(ad -bc)(il-jk)      & e(ad -bc)(il-jk)      & 0      & 0      & 0      & 0     \\
    0      & 0      & 0      & 0      & l(ad -bc)(eh-gf)     & -k(ad -bc)(eh-gf)     \\
    0      & 0      & 0      & 0      & -j(ad -bc)(eh-gf)      & i(ad -bc)(eh-gf) 
\end{array}\right)$$



---

---

## Problem

### Another decomposition

In this part, we construct an algorithm to decompose a matrix $A \in \mathbb{R}^{N\times N}$ under the form 

$$ A = Q H Q^T, \qquad (1)$$

where $Q \in \mathbb{R}^{N\times N}$ is orthogonal (i.e. $Q^{-1} = Q^T$) and $H$ satisfies $H_{i,j} = 0$ for all $i>j+1$, i.e. all the coefficients below the subdiagonal are 0 or equivalently it is of the form

$$ H = \left( \begin{array}{cccccc} 
H_{1,1} & H_{1,2} & \dots  & \dots     & H_{1,N} \\ 
H_{2,1} & H_{2,2} &        &           & H_{2,N} \\ 
0       & H_{3,2} & \ddots &           & H_{3,N} \\ 
\vdots  & \ddots  & \ddots & \ddots    & \vdots \\ 
0       & \dots   & 0      & H_{N,N-1} & H_{N,N} \\ 
\end{array}\right).$$

1) Let us denote $q^i$ the i-th column of $Q$. What is the value of the dot product $(q^i)^T (q^j)$ for $i\neq j$ and for $i = j$? 

**Answer:**

Since $Q$ is orthogonal, $Q^T Q = I$.

Therefore, $(q^i)^T (q^j) = 1$ for $i = j$ and $(q^i)^T (q^j) = 0$ for $i \neq j$

2) **Computation of the first two columns of Q and the first column of H:**

Since $Q$ is orthogonal, finding $Q$ and $H$ satisfying (1) is equivalent to finding them satisfying 

$$A Q = Q H. \qquad (2)$$ 

Let us choose $q^1 \in \mathbb{R}^N$ such that $\|q^1\| = 1$. We seek a second vector $q^2$ orthogonal to $q^1$ and such that $\|q^2\| = 1$. 

a) Using Equation (2), express $A q^1$ as a function of $q^1$, $q^2$ and the coefficients $H_{i,j}$. 

b) Since we seek a vector $q^2$ orthogonal $q^1$, deduce the value of $H_{1,1}$.  

c) Provide a vector $\tilde{q^2}$ orthogonal to $q^1$  Express it as a function of $A$, $q^1$ and $H_{1,1}$.

d) Finally, provide a vector $q^2$ that satisfies the condition in 1) and a coefficient $H_{2,1}$. 

**Answer:** 

a) $Aq^1$ is the first column of the matrix (AQ). It is then equal to the first column of $QH$. 

Then $Aq^1 = H_{1,1}q^1 + H_{2,1}q^2$.

b) We multiply both side of the previous equation by $(q^1)^T$, and get $H_{1,1} = (q^1)^TAq^1$

c) We solve that $q^2 =\frac{A-H{1,1}I}{H{2,1}}q^1.$ We have $\tilde{q^2} =(A-H{1,1}I)q_1$ which is orthogonal to $q^1$. 

d) To satisfy (1), we need to normalize $\tilde{q^2}$. Therefore we let $H{2,1} = ||\tilde{q^2}||$, and get the required $q^2 =\frac{(A-H{1,1}I)}{\|(A-A{1,1}I)q_1\|}q^1$.

3) **Iteration: Computation of the $k+1$-th column of Q and the $k$-th column of H:**

Suppose that we know the $k$ first columns $(q^i)_{i=1,\dots,k}$ of the orthogonal matrix $Q$ and $k-1$ first columns of $H$.  
 
a) Using Equation (2), express $A q^k$ as a function of the columns $q^i$ for $i = 1, \dots, k+1$ and the coefficients $H_{i,j}$ for $i=1,\dots, k$. 

b) Since we seek a vector $q^{k+1}$ orthogonal to all $q^i$ for $i=1,\dots,k$, deduce the value of $H_{i,k}$.  

c) Provide a vector $\tilde{q^{k+1}}$ orthogonal to $(q^i)_{i=1,\dots,k}$. Express it as a function of $A$, $q^i$ for $i=1,\dots,k$ and $H_{i,j}$ 

d) Finally, provide a vector $q^{k+1}$ that satisfies the condition in 1) and a coefficient $H_{k+1,k}$. 

**Remark** that at the last step, when $k=N$, then $H_{k+1,k}$ is out of the matrix.  

**Answer:**

a)$Aq^k$ is the kth column of the matrix (AQ). It is then equal to the kth column of $QH$. 

Then $Aq^k = \sum_{i=1}^{k+1} H_{i,k}q^i$.

b) We multiply both side of the previous equation by $(q^k)^T$, and get $H_{i,k} = (q^k)^TAq^k \quad \forall i = 1, .., k$

c) We solve that $q^{k+1} =\frac{(A-H{k,1}I)q^k - \sum_{i=1}^{k-1} H_{i,k}q^i)}{H{k+1,1}}.$ We have $\tilde{q^2} =(A-H{k,1}I)q^k - \sum_{i=1}^{k-1} H_{i,k}q^i)$ which is orthogonal to $(q^i)_{i=1,\dots,k}$. 

d) To satisfy (1), we need to normalize $\tilde{q^{k+1}}$. Therefore we let $H{2,1} = ||\tilde{q^{k+1}}||$, and get the required $q^{k+1}$. 


### Implementation

In this part, we assume that no forbidden operations occure. 

1) In the test below, we will use the following matrix

$$A = \left(\begin{array}{ccc} 19 & -4 & 1 \\ -1 & -5 & -1 \\ -5 & -7 & 13 \end{array}\right), $$

with the first column $q^1 = \frac{1}{3}\left(\begin{array}{c}2 \\ 1 \\ 2 \end{array}\right)$.

With these parameters, compute :

a) $A q^1$

b) $H_{1,1}$

c) $\tilde{q^2}$

d) $H_{2,1}$

e) $q^2$

*Hints:* you are allowed to use python to perform matrix products. 

In [5]:
### you may use this code cell

#initialization 
A = np.array([[19, -4, 1], [-1, -5, -1], [-5, -7, 13]])
q1 = np.array([2/3, 1/3, 2/3])
n = len(A)

#a
Aq1 = np.dot(A, q1)
print("Aq1 = ", Aq1)

#b
H11 = Aq1[0]/q1[0]
print("H11 = ", H11)

#c
tilq2 = np.dot(Aq1 - H11*np.eye(n), q1)
print("tilq2 = ", tilq2)

#d
H21 = np.linalg.norm(tilq2)
print("H21 = ", H21)

#e
q2 = tilq2/H21
print("q2 = ", q2)


Aq1 =  [12. -3.  3.]
H11 =  18.0
tilq2 =  [-3.  3. -3.]
H21 =  5.196152422706632
q2 =  [-0.57735027  0.57735027 -0.57735027]


**Answer:**

a) $A q^1 = (12, -3, 3)$

b) $H_{1,1} = 18$

c) $\tilde{q^2} = (-3, 3, -3)$

d) $H_{2,1} = 5.196152422706632$

e) $q^2 = (-0.57735027  0.57735027 -0.57735027)$


2) a) Implement a function that performs the algorithm of the last section and returns the matrices $H$ and $Q$ such that $H$. 

*Hints:* Since $H_{N+1,N}$ is out of the matrix $H$, the last step (computation of the $H_{i,N}$) needs to be computed out of the loop. 

b) Test it with the matrix in 1) and verify that
- you obtain the same values as in 1). 
- the matrix $Q$ is orthogonal.
- the matrix $H$ has the desired shape (some coefficients should be zero).
- the resulting matrix satisfies $AQ = Q H$.

In [10]:
# a)
def QTHQ_decomp(A, q1):
    """
    Computes the matrices H and Q with the previous algorithm 
    ----------   
    parameters:
    A      : matrix in the equation AQ = QH (numpy array of size N,N)
    q1     : 1st column vector in Q (normalized or not! ; numpy array of size N)
    
    returns:
    H      : matrix in the decomposition AQ = QH such that H_{i,j} = 0 for i>j+1 (numpy array of size N,N)
    Q      : orthogonal matrix in the decomposition AQ = QH (numpy array of size N,N)
    """
    
    Q      = np.zeros(np.shape(A))
    H      = np.zeros(np.shape(A))
    n = len(A)
   
    Q[0] = q1/ np.linalg.norm(q1)
    
    
    
    for k in range(n):
        Aq = np.dot(A, Q[k])
        for i in range(k):
            H[k, i] = np.dot(Q[k].T, np.dot(A,Q[k]))
        

        tilq2 = np.dot(Aq[k] - H[k, k]*np.eye(n), Q[k]) - sum([H[k,i]*Q[i] for i in range(k)])
        
        if k+1<n:
    
            H[k+1,k] = np.linalg.norm(tilq2)

            Q[k+1] = tilq2/H[k+1,k]

    
        
    return H, Q

In [11]:
# b)
### Test your algorithm here
A = np.array([[19, -4, 1], [-1, -5, -1], [-5, -7, 13]])
q1 = np.array([2/3, 1/3, 2/3])

H, Q = QTHQ_decomp(A, q1)
print("H:", H)
print("Q:", Q)
print("A:", A)
print("QH:", Q @ H)

H: [[0.         0.         0.        ]
 [3.33333333 0.         0.        ]
 [4.57842675 4.57842675 0.        ]]
Q: [[ 0.66666667  0.33333333  0.66666667]
 [ 0.57735027  0.57735027  0.57735027]
 [-0.6004569  -0.5281127  -0.6004569 ]]
A: [[19 -4  1]
 [-1 -5 -1]
 [-5 -7 13]]
QH: [[ 4.16339561  3.0522845   0.        ]
 [ 4.56785682  2.64335592  0.        ]
 [-4.50952359 -2.74914794  0.        ]]


**Final remark:** This technique is based on multiplications by orthogonal matrix. It offers good numerical stability, e.g. because such multiplications preserve the norm. It is therefore a key element of some of the most popular modern linear solvers.